### PR TITLE
util: Add circular buffer library

### PR DESCRIPTION
**Pull request by mchindri**
This library will be used in the wifi driver.
It is useful to write data into the buffer from one task and read
it from another task.
This will also be useful for other future applications or drivers

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

### CODE REVIEW

**Comment by mchindri** [comment]
v2:
 - fix codacy

**Comment by mchindri** [comment]
v3:
- fix codacy again

File: util/circular_buffer.c [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+{
+	struct circular_buffer	*ldesc;
+	uint32_t		size;
+
+	if (!desc || !element_size || !nb_elements)
+		return FAILURE;
+
+	size = nb_elements * element_size;
+	if (size < nb_elements || size < element_size)
+		return FAILURE; //Integer overflow
```
Comment (buha) [inline util/circular_buffer.c]: I am not convinced that this catches all scenarios. Somehow I have a gut feeling there is a counterexample that may get undetected by this check.
Worth checking: https://stackoverflow.com/questions/1815367/catch-and-compute-overflow-during-multiplication-of-two-large-integers

File: util/circular_buffer.c [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	int8_t		*write_ptr;
+	/** Set when the buffer is full */
+	bool		full;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+int32_t circular_buffer_init(circular_buffer_t *desc, uint32_t nb_elements,
```
Comment (buha) [inline util/circular_buffer.c]: too much to type, use `cb_init` or something like that which is shorter

File: util/circular_buffer.c [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+	if (!ldesc)
+		return FAILURE;
+
+	*nb_elements =  size_to_elements(ldesc, get_unread_size(ldesc));
+
+	return SUCCESS;
+}
+
+int32_t circular_buffer_write(circular_buffer_t *desc, void *data,
```
Comment (buha) [inline util/circular_buffer.c]: const void * data, to show it is an input buffer that is not edited

File: util/circular_buffer.c (hunk 1 above)
Comment (mchindri) [reply util/circular_buffer.c]: nb_elemnts and elemnt_size are unsigned so I think there is no counterexample. If one of them is 0 is checked in the previous if.

**Comment by mchindri** [comment]
v4:
- Replace typedef void *circular_buffer -> struct circular_buffer
- Rename functions from circular_buffer -> cb
- In cb_write replace void *data -> const void *data

**Comment by mchindri** [comment]
v5: 
 - Remove unnecessary functions. (size_to_element, element_to_size)

File: util/circular_buffer.c [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+		return FAILURE;
+	desc->nb_elem += nb_elements;
+
+	size_to_end = desc->size - ptr_to_offset(desc, desc->write_ptr);
+	if (to_write_size > size_to_end) {
+		memcpy(desc->write_ptr, data, size_to_end);
+		to_write_size -= size_to_end;
+		memcpy(desc->buff, (uint8_t *)data + size_to_end,
+		       to_write_size);
+		desc->write_ptr = desc->buff + to_write_size;
```
Comment (adrimbarean) [inline util/circular_buffer.c]: What if the write would pass the read pointer? Do we lose that data or guard against that?

**Comment by mchindri** [comment]
Todo: overwrite when buffer is full

File: util/circular_buffer.c [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+	*nb_elements =  desc->nb_elem;
+
+	return SUCCESS;
+}
+
+int32_t cb_write(struct circular_buffer *desc, const void *data,
+		 uint32_t nb_elements)
+{
+	uint32_t		to_write_size;
+	uint32_t		size_to_end;
```
Comment (mchindri) [inline util/circular_buffer.c]: remove extra tabs

**Comment by mchindri** [comment]
v6
 - Remove extra tabs
 - Enable overwritting

**Comment by mchindri** [comment]
v7 
 - Fix coday

File: util/circular_buffer.c [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	int8_t		*read_ptr;
+	/** Pointer to writing location */
+	int8_t		*write_ptr;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+int32_t cb_init(struct circular_buffer **desc, uint32_t nb_elements,
```
Comment (adrimbarean) [inline util/circular_buffer.c]: Could we get just the short description from `@brief` here in the .c file too without breaking doxygen? I am accustomed to have a short description wherever I look at the function, but maybe that is just me.

Comment (mchindri) [reply util/circular_buffer.c]: I think will be a good practice to do not write documentation in 2 places, just for maintenance reasons. If something changes, the documentation has to be modified in 2 places.
Maybe for review will be useful to be in .c file. But if you want just to use the functions, you will look at the .h file. And the review is don once.

**Comment by mchindri** [comment]
v8
 - Move comments to .c file

**Comment by mchindri** [comment]
will merge if no more comments

**Pull request merged**